File: lib/scripts/s16wgetter.py
#!/usr/bin/python

""" 
    Section 16 filing downloader
    To Use: ./s16wgetter <optional cik list, space separated>         to see the wget commands
            or
            ./s16wgetter <optional cik list, space separated> | sh    to download the filings
	    cd to directory
	    clear out output folder
	    python s16wgetter.py <optional cik list, space separated> | sh
	    python s16wgetter.py 0001144354 | sh
"""
import sys
import os

# master.idx files from: ftp://ftp.sec.gov/edgar/full-index/<4 digit year>/QTR<1-4>/master.idx
master_idx_files = []
for filename in os.listdir(sys.argv[2]):
    master_idx_files.append(os.path.join(sys.argv[2], filename))
    
def pad_cik(cik):
    c = cik
    if len(c) < 10:
        padding = 10 - len(c)
        new_c = ''
        while padding > 0:
            new_c = new_c + "0"
            padding = padding - 1
        new_c = new_c + c
        c = new_c
    return c

import commands
for file in master_idx_files:
    f = open(file)
    lines = f.readlines()
    for line in lines:
        if len(line.split('|')) != 5:
            continue
        form_type = line.split('|')[2]
        type = form_type.replace('/A', '')
        if type != '3' and type != '4' and type != '5':
            continue

        cik = pad_cik(line.split('|')[0])
        if len(sys.argv[2:]) > 1:
            found_cik = False
            for cik_arg in sys.argv[2:]:
                if pad_cik(cik_arg) == cik:
                    found_cik = True
                    break
            if not found_cik:
                continue

        path = line.split('|')[4]
        accession_number = path.split('/')[3].replace('.txt', '').strip()
        split_path = path.split('/')
        outputDir = sys.argv[1]

        url = "www.sec.gov/%s/%s/%s/%s/edgar.xml -O %s/%s.xml" % (
        split_path[0], split_path[1], split_path[2], accession_number.replace('-', ''), outputDir, accession_number)
        print url
        # os.system(url)
        print

        url = "wget -nv https://www.sec.gov/Archives/%s/%s/%s/%s/edgar.xml -O %s/%s.xml" % (
        split_path[0], split_path[1], split_path[2], accession_number.replace('-', ''), outputDir, accession_number)
        commands.getstatusoutput(url)
        print url
        os.system(url)
        print
if f:
    f.close()
print "finished"
sys.stdout.flush()

File: lib/scripts/s16wgetter.js
const readline = require('readline');
const fs = require('fs');
const https = require('https');
const exec = require('child_process').exec;
const wget = require('wget-improved');

/**
 * Section 16 filing downloader.
 * 
 * @param {string} outputDir directory to output files in
 * @param {string} midxDir directory to find master.idx files in
 * @param {array} cikList list of CIK numbers to filter by
 */
async function s16wgetter(outputDir, midxDir, cikList, progress) {
    let master_idx_files = await readdir(midxDir);
    let paddedCikList = [];
    for (let i = 0; i < cikList.length; ++i) {
        paddedCikList.push(padCik(cikList[i]));
    }
    let filePromises = [];
    for(let i = 0; i < master_idx_files.length; ++i) {
        let promise = new Promise((resolve, reject) => {
            let filePath = `${midxDir}/${master_idx_files[i]}`;
            let lineReader = readline.createInterface({
                input: fs.createReadStream(filePath)
            });
            lineReader.on('line', (line) => {
                lineReader.pause();
                handleMasterIDXLine(line, outputDir, paddedCikList, (prog, tot) => {
                    if (progress) progress(master_idx_files[i], prog, tot);
                }).then(() => {
                    lineReader.resume();
                });
            });
            lineReader.on('close', function () {
                resolve();
            });
        });
        filePromises.push(promise);
    }
    await Promise.all(filePromises);
    return;
}

/**
 * Parse line from master.idx file and download corresponding file(s)
 * to the outputDir.
 * 
 * @param {string} line line being read
 * @param {string} outputDir directory to output files in
 * @param {array} cikList list of CIK numbers to filter by
 */
async function handleMasterIDXLine(line, outputDir, cikList, progress) {
    let splitLine = line.split('|');
    if (splitLine.length != 5) return;

    let formType = splitLine[2];
    let type = formType.replace('/A', '');
    if (['3', '4', '5'].indexOf(type) < 0) return;

    let cik = padCik(splitLine[0]);
    if (cikList.indexOf(cik) == -1) return;

    let path = splitLine[4];
    let splitPath = path.split('/');
    let accessionNumber = splitPath[3].replace('.txt', '').trim();
    let url = `https://www.sec.gov/Archives/${splitPath[0]}/${splitPath[1]}/${splitPath[2]}/${accessionNumber.replace(/-/g, '')}/edgar.xml`;
    // download to `${outputDir}/${accessionNumber}.xml`
    await downloadFile(url, `${outputDir}/${accessionNumber}.xml`, progress);
}

/**
 * Get files in a directory at path.
 * 
 * @param {string} path 
 */
function readdir(path) {
    return new Promise((res, rej) => {
        console.log(path);
        fs.readdir(path, (err, files) => {
            if (err) {
                rej(err);
                return;
            }
            res(files);
        })
    });
}

/**
 * ¯\_(ツ)_/¯ copy pasta from s16wgetter.py
 * 
 * @param {string} cik 
 */
function padCik(cik) {
    let c = cik;
    if (c.length < 10) {
        let padding = 10 - c.length;
        let new_c = '';
        for (;padding > 0; --padding) {
            new_c += '0';
        }
        new_c += c;
        c = new_c;
    }
    return c;
}

async function downloadFile(url, outputPath, progress) {
    let download = wget.download(url, outputPath, {

    });
    download.on('error', (err) => {
        console.error(url);
        console.error(err);
    });
    if (progress)
        download.on('progress', prog => {
            // console.log(prog);
            // progress(prog, 1);
        });
}

module.exports = s16wgetter;